File: src/leetcode/arrays_hashing/TwoSum.jl
"""
    two_sum(nums::Vector{Int}, target::Int) -> Vector{Int}

Given a vector of integers `nums` and an integer `target`, return the 1-based indices
of the two numbers such that they add up to `target`.

Assumes that each input has exactly one solution, and the same element cannot be used twice.
Throws an error if no solution is found.

# Arguments
- `nums::Vector{Int}`: The input vector of integers.
- `target::Int`: The target sum.

# Returns
- `Vector{Int}`: A vector containing the two 1-based indices.

# Throws
- `ErrorException`: If no two numbers in `nums` sum to `target`.

# Examples
```jldoctest
julia> twoSum([2, 7, 11, 15], 9)
2-element Vector{Int64}:
 1
 2

julia> twoSum([3, 2, 4], 6)
2-element Vector{Int64}:
 2
 3
```
"""
function two_sum(nums::Vector{Int}, target::Int)::Vector{Int}
    num_to_index = Dict{Int,Int}() # Stores number -> index mapping
    for (index, num) in enumerate(nums)
        complement = target - num
        # Use get with default 0 (invalid index) to check if complement exists
        complement_index = get(num_to_index, complement, 0)
        if complement_index > 0 # Check if a valid index was found
            return [complement_index, index]
        end
        # Store the current number and its index
        num_to_index[num] = index
    end
    # If the loop completes, no solution was found (violates assumption)
    error("No two sum solution found for the given input.")
end

File: src/leetcode/rust/src/arrays_hashing/two_sum.rs
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    /// Find two indices in nums such that they add up to target.
    ///
    /// Uses a HashMap to track seen values and their indices for O(n) time complexity.
    ///
    /// # Arguments
    /// * `nums` - Vector of integers
    /// * `target` - Target sum
    ///
    /// # Returns
    /// Vector containing two indices [i, j] where nums[i] + nums[j] == target.
    /// Returns empty vector if no solution exists.
    ///
    /// # Examples
    /// ```
    /// use leetcode_rust::arrays_hashing::two_sum::Solution;
    ///
    /// let result = Solution::two_sum(vec![2, 7, 11, 15], 9);
    /// assert_eq!(result, vec![0, 1]);
    /// ```
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<usize> {
        let mut seen: HashMap<i32, usize> = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            let difference = target - num;

            if let Some(&j) = seen.get(&difference) {
                return vec![j, i];
            }

            seen.insert(num, i);
        }

        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_case() {
        assert_eq!(Solution::two_sum(vec![2, 7, 11, 15], 9), vec![0, 1]);
    }

    #[test]
    fn test_middle_indices() {
        assert_eq!(Solution::two_sum(vec![3, 2, 4], 6), vec![1, 2]);
    }

    #[test]
    fn test_duplicate_values() {
        assert_eq!(Solution::two_sum(vec![3, 3], 6), vec![0, 1]);
    }

    #[test]
    fn test_larger_array() {
        assert_eq!(Solution::two_sum(vec![1, 2, 3, 4, 5], 9), vec![3, 4]);
    }

    #[test]
    fn test_no_solution() {
        assert_eq!(Solution::two_sum(vec![1, 2, 3], 10), vec![]);
    }

    #[test]
    fn test_negative_numbers() {
        assert_eq!(Solution::two_sum(vec![-1, -2, -3, -4, -5], -8), vec![2, 4]);
    }
}
